### PR TITLE
applets: Add AppletManager and implement PhotoViewer and Error applets

### PR DESCRIPTION
**Pull request by DarkLordZach**
This does a couple of things:
- Adds a centralized `AppletManager` class whose sole purpose is to manage lifetime for the frontends of applets. They can be set/cleared and when requested passed via an applet backend. This is a much better approach than adding a getter/setter/const getter triad every applet in core.
- Makes all applets take the frontend as a const reference in the constructor (as opposed to using `Core::System::GetInstance()`) this consolidates the calls to `GetInstance()` which should make it much easier in the future to deglobalize.
- Implements the Error applet fully: This applet is the error display screen on hardware and can display simple error codes, errors with timestamps, and custom app-defined messages. On the frontend, this is done with dialogs which are pictured below.
- Implements the PhotoViewer applet partially: Only adds the backend (which will parse the request) and the Default frontend impl. (which logs that the action occured), adding a qt frontend for this seemed unnecessary at this time, however all it would require is a subclass of `Core::Frontend::PhotoViewerApplet`
- Moves `StubApplet` to `general_backend.[cpp|h]`. Since there are a variety of applets (i.e. PhotoViewer, stub) that are almost trivially implemented, it doesn't make sense to have a cpp/h pair for each. `general_[back|front]end.[cpp|h]` consolidates these into one pair.
- The Error applet and PhotoViewer applet details are based on RE work I did in IDA, which I then verified with 2 hardware tests. I will post these on my GitHub as soon as i can.

Sample of Error Applet Frontend:
error1
error2
error3

Some of this code was written before I went on hiatus so if i did something egregious lmk & sorry.

### CODE REVIEW

File: src/core/hle/service/am/applets/error.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    ShowApplicationError = 2,
+    ShowEula = 3,
+    ShowErrorPctl = 4,
+    ShowErrorRecord = 5,
+    ShowUpdateEula = 8,
+};
+
+class Error final : public Applet {
+public:
+    Error(const Core::Frontend::ErrorApplet& frontend);
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.h]:
```suggestion
    explicit Error(const Core::Frontend::ErrorApplet& frontend);
```

File: src/core/hle/service/am/applets/error.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "core/hle/result.h"
+#include "core/hle/service/am/applets/applets.h"
+
+namespace Service::AM::Applets {
+
+union ErrorArguments;
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.h]: This forward declaration can be moved into the private section of the Error class.

File: src/core/hle/service/am/applets/general_backend.cpp [hunk 3]
```diff
@@ -30,6 +33,55 @@ static void LogCurrentStorage(AppletDataBroker& broker, std::string prefix) {
...
+PhotoViewer::PhotoViewer(const Core::Frontend::PhotoViewerApplet& frontend) : frontend(frontend) {}
+
+PhotoViewer::~PhotoViewer() = default;
+
+void PhotoViewer::Initialize() {
+    Applet::Initialize();
+    complete = false;
+
+    const auto storage = broker.PopNormalDataToApplet();
+    ASSERT(storage != nullptr);
```
Comment (lioncash) [inline src/core/hle/service/am/applets/general_backend.cpp]: `common/assert.h` needs to be included (as well as `common/logging/log.h`

File: src/core/hle/service/am/applets/error.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    ShowError error;
+    ShowErrorRecord error_record;
+    SystemErrorArg system_error;
+    ApplicationErrorArg application_error;
+};
+
+namespace {
+template <typename T>
+void CopyArgumentData(const std::vector<u8>& data, T& variable) {
+    ASSERT(data.size() >= sizeof(T));
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.cpp]: `common/assert.h` and `common/logging/log.h` need to be included.

File: src/core/hle/service/am/applets/error.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    ResultCode GetStatus() const override;
+    void ExecuteInteractive() override;
+    void Execute() override;
+
+    void DisplayCompleted();
+
+private:
+    const Core::Frontend::ErrorApplet& frontend;
+    ResultCode error_code = RESULT_SUCCESS;
+    ErrorAppletMode mode;
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.h]: this and `complete` should have default values on construction.

File: src/core/hle/service/am/applets/general_backend.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+namespace Service::AM::Applets {
+
+enum class PhotoViewerAppletMode : u8 {
+    CurrentApp = 0,
+    AllApps = 1,
+};
+
+class PhotoViewer final : public Applet {
+public:
+    PhotoViewer(const Core::Frontend::PhotoViewerApplet& frontend);
```
Comment (lioncash) [inline src/core/hle/service/am/applets/general_backend.h]:
```suggestion
    explicit PhotoViewer(const Core::Frontend::PhotoViewerApplet& frontend);
```

File: src/core/hle/service/am/applets/general_backend.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    bool TransactionComplete() const override;
+    ResultCode GetStatus() const override;
+    void ExecuteInteractive() override;
+    void Execute() override;
+
+    void ViewFinished();
+
+private:
+    const Core::Frontend::PhotoViewerApplet& frontend;
+    bool complete;
```
Comment (lioncash) [inline src/core/hle/service/am/applets/general_backend.h]: this and mode should have default values on construction

File: src/yuzu/applets/error.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    connect(this, &QtErrorDisplay::MainWindowDisplayError, &parent,
+            &GMainWindow::ErrorDisplayDisplayError, Qt::QueuedConnection);
+    connect(&parent, &GMainWindow::ErrorDisplayFinished, this,
+            &QtErrorDisplay::MainWindowFinishedError, Qt::DirectConnection);
+}
+
+QtErrorDisplay::~QtErrorDisplay() = default;
+
+void QtErrorDisplay::ShowError(ResultCode error, std::function<void()> finished) const {
+    this->callback = finished;
```
Comment (lioncash) [inline src/yuzu/applets/error.cpp]:
```suggestion
    this->callback = std::move(finished);
```

File: src/yuzu/applets/error.cpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            .arg(QDateTime::fromSecsSinceEpoch(time.count()).toString("h:mm:ss A"))
+            .arg(static_cast<u32>(error.module.Value()) + 2000, 4, 10, QChar::fromLatin1('0'))
+            .arg(error.description, 4, 10, QChar::fromLatin1('0'))
+            .arg(error.raw, 8, 16, QChar::fromLatin1('0')));
+}
+
+void QtErrorDisplay::ShowCustomErrorText(ResultCode error, std::string dialog_text,
+                                         std::string fullscreen_text,
+                                         std::function<void()> finished) const {
+    this->callback = finished;
```
Comment (lioncash) [inline src/yuzu/applets/error.cpp]:
```suggestion
    this->callback = std::move(finished);
```

File: src/core/hle/service/am/applets/error.cpp [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    Applet::Initialize();
+    args = std::make_unique<ErrorArguments>();
+    complete = false;
+
+    const auto storage = broker.PopNormalDataToApplet();
+    ASSERT(storage != nullptr);
+    const auto data = storage->GetData();
+
+    ASSERT(!data.empty());
+    std::memcpy(&mode, data.data(), sizeof(ErrorAppletMode));
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.cpp]: `<cstring>` needs to be included.

File: src/core/hle/service/am/applets/error.cpp [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    std::array<char, 0x800> detail_text;
+};
+static_assert(sizeof(SystemErrorArg) == 0x1018, "SystemErrorArg has incorrect size.");
+
+struct ApplicationErrorArg {
+    u8 mode;
+    bool jump;
+    INSERT_PADDING_BYTES(6);
+    u32 error_code;
+    std::array<char, 8> language_code;
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.cpp]: `<array>` needs to be included.

File: src/core/hle/service/am/applets/error.cpp [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        CopyArgumentData(data, args->error);
+        if (args->error.use_64bit_error_code) {
+            error_code = Decode64BitError(args->error.error_code_64);
+        } else {
+            error_code = ResultCode(args->error.error_code_32);
+        }
+        break;
+    case ErrorAppletMode::ShowSystemError:
+        CopyArgumentData(data, args->system_error);
+        error_code = ResultCode(Decode64BitError(args->system_error.error_code_64));
```
Comment (lioncash) [inline src/core/hle/service/am/applets/error.cpp]: missing a `break`?

File: src/yuzu/applets/web_browser.h [hunk 13]
```diff
@@ -38,7 +38,7 @@ class QtWebBrowser final : public QObject, public Core::Frontend::WebBrowserAppl
     ~QtWebBrowser() override;
 
     void OpenPage(std::string_view url, std::function<void()> unpack_romfs_callback,
-                  std::function<void()> finished_callback) override;
+                  std::function<void()> finished_callback) const override;
```
Comment (lioncash) [inline src/yuzu/applets/web_browser.h]: This wasmade non-const in a661025637f301013d9da781e67334853162c6b3 for a reason.

File: src/yuzu/applets/error.cpp [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        tr("An error has occured.\nPlease try again or contact the developer of the "
+           "software.\n\nError Code: %1-%2 (0x%3)")
+            .arg(static_cast<u32>(error.module.Value()) + 2000, 4, 10, QChar::fromLatin1('0'))
+            .arg(error.description, 4, 10, QChar::fromLatin1('0'))
+            .arg(error.raw, 8, 16, QChar::fromLatin1('0')));
+}
+
+void QtErrorDisplay::ShowErrorWithTimestamp(ResultCode error, std::chrono::seconds time,
+                                            std::function<void()> finished) const {
+    this->callback = finished;
```
Comment (lioncash) [inline src/yuzu/applets/error.cpp]:
```suggestion
    this->callback = std::move(finished);
```

File: src/core/hle/service/am/applets/software_keyboard.h [hunk 15]
```diff
@@ -55,7 +55,7 @@ static_assert(sizeof(KeyboardConfig) == 0x3E0, "KeyboardConfig has incorrect siz
 
 class SoftwareKeyboard final : public Applet {
 public:
-    SoftwareKeyboard();
+    SoftwareKeyboard(const Core::Frontend::SoftwareKeyboardApplet& frontend);
```
Comment (lioncash) [inline src/core/hle/service/am/applets/software_keyboard.h]:
```suggestion
    explicit SoftwareKeyboard(const Core::Frontend::SoftwareKeyboardApplet& frontend);
```

File: src/core/hle/service/am/applets/profile_select.h [hunk 16]
```diff
@@ -28,7 +29,7 @@ static_assert(sizeof(UserSelectionOutput) == 0x18, "UserSelectionOutput has inco
 
 class ProfileSelect final : public Applet {
 public:
-    ProfileSelect();
+    ProfileSelect(const Core::Frontend::ProfileSelectApplet& frontend);
```
Comment (lioncash) [inline src/core/hle/service/am/applets/profile_select.h]:
```suggestion
    explicit ProfileSelect(const Core::Frontend::ProfileSelectApplet& frontend);
```

File: src/core/hle/service/am/applets/profile_select.h [hunk 17]
```diff
@@ -7,6 +7,7 @@
 #include <vector>
 
 #include "common/common_funcs.h"
+#include "core/frontend/applets/software_keyboard.h"
```
Comment (lioncash) [inline src/core/hle/service/am/applets/profile_select.h]: What makes use of the software keyboard in this header?

**Comment by bunnei** [comment]
Will add to canary once #1957 is merged

**Comment by bunnei** [comment]
@DarkLordZach can you rebase this?

**Pull request merged**